File: J9Ch/src/S_2_Data_Structure_I/Required_12/_432_Find_the_Weak_Connected_Component_in_the_Directed_Graph_Medium.cpp


/**
 * Definition for Directed graph.
 * struct DirectedGraphNode {
 *     int label;
 *     vector<DirectedGraphNode *> neighbors;
 *     DirectedGraphNode(int x) : label(x) {};
 * };
 */
map<int, int> f, g;
class Solution {
public:
    /**
     * @param nodes a array of directed graph node
     * @return a connected set of a directed graph
     */
    int find(int x) {
        if (x == f[x])
            return x;

        return f[x] = find(f[x]);
    }

    void merge(int x, int y) {
        x = find(x);
        y = find(y);
        if (x != y)
            f[x] = y;
    }

    vector<vector<int>> connectedSet2(vector<DirectedGraphNode*>& nodes) {
        // Write your code here
        vector<vector<int>> result;
        int n = nodes.size();
        for (int i = 0; i < n; ++i) f[nodes[i]->label] = nodes[i]->label;

        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < nodes[i]->neighbors.size(); ++j) {
                DirectedGraphNode* tmp = nodes[i]->neighbors[j];
                merge(nodes[i]->label, tmp->label);
            }
        }
        int cnt = 0;
        for (int i = 0; i < n; ++i)
        {
            int x = find(nodes[i]->label);
            if (g.find(x) == g.end())
                g[x] = ++cnt ;

            if (result.size() < cnt) {
                vector<int> tmp;
                result.push_back(tmp);
            }
            result[g[x]-1].push_back(nodes[i]->label);
        }
        return result;
    }
};

File: J9Ch/src/S_2_Data_Structure_I/Required_12/_432_Find_the_Weak_Connected_Component_in_the_Directed_Graph_Medium.py


# Definition for a directed graph node
# class DirectedGraphNode:
#     def __init__(self, x):
#         self.label = x
#         self.neighbors = []
class Solution:
    # @param {DirectedGraphNode[]} nodes a array of directed graph node
    # @return {int[][]} a connected set of a directed graph
    def __init__(self):
        self.f = {}

    def merge(self, x, y):
        x = self.find(x)
        y = self.find(y)
        if x != y:
            self.f[x] = y

    def find(self, x):
        if self.f[x] == x:
            return x

        self.f[x] = self.find(self.f[x])
        return self.f[x]

    def connectedSet2(self, nodes):
        # Write your code here
        for node in nodes:
            self.f[node.label] = node.label

        for node in nodes:
            for nei in node.neighbors:
                self.merge(node.label, nei.label)

        result = []
        g = {}
        cnt = 0

        for node in nodes:
            x = self.find(node.label)
            if not x in g:
                cnt += 1
                g[x] = cnt

            if len(result) < cnt:
                result.append([])

            result[g[x] - 1].append(node.label)

        return result